### PR TITLE
[swiftsrc2cpg] Fix subscript handling with nested accessor methods

### PR DESCRIPTION
**Pull request by max-leuthaeuser**
- does not emit a separate METHOD node for subscript functions if they contain nested accessors anymore
- resolves closure binding issues there (identifiers ref to subscript function parameters correctly now)

### CODE REVIEW

File: joern-cli/frontends/swiftsrc2cpg/src/main/scala/io/joern/swiftsrc2cpg/astcreation/AstCreatorHelper.scala [hunk 1]
```diff
@@ -347,18 +338,30 @@ trait AstCreatorHelper(implicit withSchemaValidation: ValidationMode) { this: As
...
-  protected def methodInfoForAccessorDecl(node: AccessorDeclSyntax, variableName: String, tpe: String): MethodInfo = {
+  protected def methodInfoForAccessorDecl(
+    node: AccessorDeclSyntax,
+    variableName: String,
+    tpe: String,
+    fullNameSubscriptPrefix: String = ""
+  ): MethodInfo = {
     val accessorName = code(node.accessorSpecifier)
+    val namePrefix = if (variableName.nonEmpty) { s"$variableName." }
+    else { "" }
```
Comment (maltek) [inline joern-cli/frontends/swiftsrc2cpg/src/main/scala/io/joern/swiftsrc2cpg/astcreation/AstCreatorHelper.scala]: This style of `if/else` that scalafmt sometimes spits out where a `else` in expression context doesn't get any indentation just doesn't parse in my head. I seem to see the `else` first, and then search around where I've missed the `if`.

I always manually edit them like this (which scalamft won't undo):
```suggestion
    val namePrefix =
      if (variableName.nonEmpty) { s"$variableName." }
      else { "" }
```

or this (which might be undone by `scalafmt` if the line is too long):
```suggestion
    val namePrefix = if (variableName.nonEmpty) s"$variableName."  else ""
```

It's still just a minor style thing so do as you wish with that.

File: joern-cli/frontends/swiftsrc2cpg/src/main/scala/io/joern/swiftsrc2cpg/astcreation/AstCreatorHelper.scala [hunk 2]
```diff
@@ -373,7 +376,12 @@ trait AstCreatorHelper(implicit withSchemaValidation: ValidationMode) { this: As
       case None =>
         val (methodName, methodFullName) = calcNameAndFullName(name)
         registerType(tpe)
-        MethodInfo(methodName, methodFullName, tpe, tpe)
+        val methodFullNameWithSubscriptPrefix = if (fullNameSubscriptPrefix.nonEmpty) {
+          methodFullName.replaceFirst(s"\\.$methodName", s".$fullNameSubscriptPrefix.$methodName")
```
Comment (maltek) [inline joern-cli/frontends/swiftsrc2cpg/src/main/scala/io/joern/swiftsrc2cpg/astcreation/AstCreatorHelper.scala]: could we build the fullname correctly in `calcNameAndFullName` to begin with? These kinds of string replacements have a tendency to break in weird edge cases.

File: joern-cli/frontends/swiftsrc2cpg/src/test/scala/io/joern/swiftsrc2cpg/passes/ast/ExtensionTests.scala [hunk 3]
```diff
@@ -91,65 +90,92 @@ class ExtensionTests extends SwiftSrc2CpgSuite {
...
+          |      return path.reduce(self) { $0[sub: $1] }
+          |    }
+          |  }
+          |
+          |  public subscript(path: JSONSubscriptType...) -> JSON {
+          |    get {
+          |      return self[path]
+          |    }
+          |  }
+          |}""".stripMargin
```
Comment (maltek) [inline joern-cli/frontends/swiftsrc2cpg/src/test/scala/io/joern/swiftsrc2cpg/passes/ast/ExtensionTests.scala:151]: it would be good to also have a test with a read-only subscript without the explicit `get` block (3rd code sample at https://docs.swift.org/swift-book/documentation/the-swift-programming-language/subscripts/#Subscript-Syntax )

File: joern-cli/frontends/swiftsrc2cpg/src/test/scala/io/joern/swiftsrc2cpg/passes/ast/ExtensionTests.scala [hunk 4]
```diff
@@ -91,65 +90,92 @@ class ExtensionTests extends SwiftSrc2CpgSuite {
...
-      val List(getRef1)                  = subscriptMethod1.ast.isMethodRef.code("get").l
-      val List(closureBindingPath1)      = getRef1.captureOut.l
-      closureBindingPath1.closureBindingId shouldBe Option(
-        "Ext1.swift:<global>.Foo.subscript:(path:Swift.Array)->JSON:get:ANY:path"
-      )
-      closureBindingPath1.refOut.head shouldBe pathParam1
-      closureBindingPath1.evaluationStrategy shouldBe EvaluationStrategies.BY_REFERENCE
-
-      inside(cpg.call.codeExact("self[path]").argument(2).l) { case List(id: Identifier) =>
+      inside(subscriptFooGet2.ast.isCall.codeExact("self[path]").argument(2).l) { case List(id: Identifier) =>
```
Comment (maltek) [inline joern-cli/frontends/swiftsrc2cpg/src/test/scala/io/joern/swiftsrc2cpg/passes/ast/ExtensionTests.scala:199]: from the argument index 2 I gather this is still lowered as a index access, not as a call to the overloaded subscript method? That's fine for now, but another todo :)

File: joern-cli/frontends/swiftsrc2cpg/src/main/scala/io/joern/swiftsrc2cpg/astcreation/AstForDeclSyntaxCreator.scala [hunk 5]
```diff
@@ -595,15 +595,55 @@ trait AstForDeclSyntaxCreator(implicit withSchemaValidation: ValidationMode) {
...
+    }
+  }
+
+  private def astForAccessorInSubscript(
+    node: SubscriptDeclSyntax,
+    handleAccessor: (AccessorDeclSyntax, String, String, String, Seq[SwiftNode]) => Unit
+  ): Ast = {
+    val methodInfo                                       = methodInfoForFunctionDeclLike(node)
+    val MethodInfo(methodName, _, signature, returnType) = methodInfo
+    val subscriptSignaturePrefix                         = s"$methodName:${signature.replace(s"->$returnType", "")}"
```
Comment (maltek) [inline joern-cli/frontends/swiftsrc2cpg/src/main/scala/io/joern/swiftsrc2cpg/astcreation/AstForDeclSyntaxCreator.scala]: similar concern here about the string replacement. Here, I'm pretty sure it would break when some function pointers are involved. (Even if I don't know why someone would implement a subscript that takes functions as arguments.)

File: joern-cli/frontends/swiftsrc2cpg/src/test/scala/io/joern/swiftsrc2cpg/passes/ast/ExtensionTests.scala (hunk 4 above)
Comment (max-leuthaeuser) [reply joern-cli/frontends/swiftsrc2cpg/src/test/scala/io/joern/swiftsrc2cpg/passes/ast/ExtensionTests.scala:199]: Yes, that's for another PR.
